### PR TITLE
Versioning library

### PR DESCRIPTION
**Pull request by johndebord**
## Change Description

In reference to [Issue #7396](https://github.com/EOSIO/eos/issues/7396) and previous [Pull Request #7414](https://github.com/EOSIO/eos/pull/7414), this PR alters the format of printing the `cleos` version in a more intuitive way by the addition of a new library in repository `EOSIO/eos`— `version`. This library is extensible enough to be included into other programs as well. Also, a new `cleos` sub-command has been added as well; `cleos version full`.

<hr>

**Under The Hood:**
We can see that when running `make`, the build system updates the version metadata if need be.
```shell
~/eos/build $ make
.
.
.
[ 83%] Built target eosio_chain
[ 84%] Built target eosio_testing
[ 84%] Updating version metadata... <---
[ 84%] Built target evaluate_every_build <---
[ 85%] Built target version <---
[ 86%] Built target chain_plugin
[ 87%] Built target http_client_plugin
.
.
.
```
This notable behavior can be found in file `/eos/libraries/version/CMakeLists.txt`.
```cmake
.
.
.
# Create a custom target to update the version metadata upon every build.
find_package(Git)
if(EXISTS ${CMAKE_SOURCE_DIR}/.git AND ${GIT_FOUND})
   add_custom_target(
      evaluate_every_build ALL
      COMMAND ${CMAKE_COMMAND} -DGIT_EXEC=${GIT_EXECUTABLE}
                               -DCUR_BIN_DIR=${CMAKE_CURRENT_BINARY_DIR}
                               -DCUR_SRC_DIR=${CMAKE_CURRENT_SOURCE_DIR}
                               -DSRC_DIR=${CMAKE_SOURCE_DIR}
                               -DV_MAJOR=${VERSION_MAJOR}
                               -DV_MINOR=${VERSION_MINOR}
                               -DV_PATCH=${VERSION_PATCH}
                               -DV_SUFFIX=${VERSION_SUFFIX}
                               -P ${CMAKE_SOURCE_DIR}/CMakeModules/VersionUtils.cmake
      COMMENT "Updating version metadata..." VERBATIM )
else()
   # Modify and substitute the `.cpp.in` file for a `.cpp` in the build directory.
   configure_file(
      ${CMAKE_CURRENT_SOURCE_DIR}/src/version_impl.cpp.in
      ${CMAKE_CURRENT_BINARY_DIR}/src/version_impl.cpp
      @ONLY )
endif()
.
.
.
```

<hr>

**Examples:**
```shell
~/eos/build/bin $ ./cleos version client
1.9.0-develop
```

```shell
~/eos/build/bin $ ./cleos version full
v1.9.0-develop-cccfb5cd4a1d98b8b4392c2969713bd13b9eee8b
```

<hr>

**Walk-through:**
```shell
~/eos/build/bin $ git describe --tags --dirty; ./nodeos --version; ./cleos version full
v1.8.0-rc1-590-gcccfb5cd4
v1.8.0-rc1-590-gcccfb5cd4
v1.9.0-develop-cccfb5cd4a1d98b8b4392c2969713bd13b9eee8b
```

```shell
~/eos/build/bin $ echo "//yo" >> ../../programs/cleos/main.cpp 
~/eos/build/bin $ git describe --tags --dirty; ./nodeos --version; ./cleos version full
v1.8.0-rc1-590-gcccfb5cd4-dirty
v1.8.0-rc1-590-gcccfb5cd4
v1.9.0-develop-cccfb5cd4a1d98b8b4392c2969713bd13b9eee8b
```

```shell
~/eos/build $ make -j20
.
.
.
[100%] Built target nodeos
[100%] Linking CXX executable cleos
[100%] Built target cleos
```

```shell
~/eos/build/bin $ git describe --tags --dirty; ./nodeos --version; ./cleos version full
v1.8.0-rc1-590-gcccfb5cd4-dirty
v1.8.0-rc1-590-gcccfb5cd4-dirty
v1.9.0-develop-cccfb5cd4a1d98b8b4392c2969713bd13b9eee8b-dirty
```

```shell
~/eos/build/bin $ git status
On branch version-library
Your branch is up to date with 'origin/version-library'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   ../../programs/cleos/main.cpp

no changes added to commit (use "git add" and/or "git commit -a")
~/eos/build/bin $ git add ../../programs/cleos/main.cpp; git commit -m "Test"
[version-library 2e89339aa] Test
 1 file changed, 1 insertion(+)
```

```shell
~/eos/build/bin $ git describe --tags --dirty; ./nodeos --version; ./cleos version full
v1.8.0-rc1-591-g2e89339aa
v1.8.0-rc1-590-gcccfb5cd4-dirty
v1.9.0-develop-cccfb5cd4a1d98b8b4392c2969713bd13b9eee8b-dirty
```

```shell
~/eos/build $ make -j20
.
.
.
[100%] Built target nodeos
[100%] Linking CXX executable cleos
[100%] Built target cleos
```

```shell
~/eos/build/bin $ git describe --tags --dirty; ./nodeos --version; ./cleos version full
v1.8.0-rc1-591-g2e89339aa
v1.8.0-rc1-591-g2e89339aa
v1.9.0-develop-2e89339aad51f26ac41de43984aa2189767845d1
```

<hr>

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [x] API Changes

New `cleos version` sub-command: `cleos version full`.

## Documentation Additions
- [x] Documentation Additions

### CODE REVIEW

File: libraries/version/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+project(
+  Version )
+
+# Found in directory `eos/CMakeModules/`.
+include(
+  VersionUtils )
+
+# Generate the most up-to-date version metadata of the repository.
+GENERATE_VERSION_METADATA()
```
Comment (spoonincode) [inline libraries/version/CMakeLists.txt]: It appears this generation is being done at cmake configure/generation and not when building. If we're going to report dirtiness we need more accuracy. To get this, will need to replicate how appbase creates a dummy target that is run on each build to populate the build information.

File: libraries/version/src/version_impl.cpp.in [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+   const std::string& _version_client() {
+      static const std::string version{'v' + version_major +
+                                       '.' + version_minor +
+                                       '.' + version_patch +
+                                       '-' + version_suffix};
+      return version;
+   }
+
+   const std::string& _version_full() {
```
Comment (spoonincode) [inline libraries/version/src/version_impl.cpp.in]: This is going to potentially give a different version than what nodeos reports (and if we modify nodeos to use the new version string it means we're going to lose some information). Is that intentional?

**Comment by NorseGaud** [comment]
Make sure to merge in develop to this branch or CICD will continue to fail

File: libraries/version/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+# Make dependencies visible.
+target_include_directories(
+   version
+   PUBLIC
+   "${CMAKE_CURRENT_SOURCE_DIR}/include/"
+   PRIVATE
+   "${CMAKE_CURRENT_SOURCE_DIR}/src/" )
+
+# Install the library in the appropriate places.
```
[the remark this answers is not in the document]
Comment (johndebord) [reply libraries/version/CMakeLists.txt]: Updated.

File: libraries/version/CMakeLists.txt (hunk 1 above)
Comment (johndebord) [reply libraries/version/CMakeLists.txt]: Updated.

File: libraries/version/src/version_impl.cpp.in (hunk 2 above)
Comment (johndebord) [reply libraries/version/src/version_impl.cpp.in]: Yes, the plan was to have this specific versioning output from now on.
This functionality is also planned to be added into `nodeos` and the other programs alike as well if everything is agreed upon.

**Comment by johndebord** [comment]
@spoonincode Very interesting. This seems to be a caveat associated with `ninja`. I'll look into it.

<hr>

```shell
~/eos $ mkdir ninja; mkdir make; pushd ./; cd ninja; cmake -GNinja ..; popd; pushd ./; cd make; cmake ..; popd;
.
.
.
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/john.debord/eos/ninja
.
.
.
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/john.debord/eos/make	
```

```shell
~/eos $ pushd ./; cd ninja; ninja; popd; pushd ./; cd make; make -j20; popd;
.
.
.
[437/601] Building CXX object libraries/chain/CMakeFiles/eosio_chain.dir/resource_limits.cpp.o
[438/601] Building CXX object libraries/chain/CMakeFiles/eosio_chain.dir/thread_utils.cpp.o
[439/601] Updating version metadata... <---
[441/601] Building CXX object libraries/wabt/CMakeFiles/wabt.dir/src/token.cc.o
[442/601] Building CXX object libraries/wabt/CMakeFiles/wabt.dir/src/opcode.cc.o
.
.
.
[  6%] Building CXX object libraries/wabt/CMakeFiles/wabt.dir/src/prebuilt/wast-lexer-gen.cc.o
[  5%] Linking C static library libbuiltins.a
/Library/Developer/CommandLineTools/usr/bin/ranlib: Updating version metadata... <---
[  5%] Built target appbase_version
[  6%] Building CXX object libraries/wabt/CMakeFiles/wabt.dir/src/lexer-source.cc.o
.
.
.
```

```shell
~/eos $ ./ninja/bin/cleos version full; ./make/bin/cleos version full;
v1.9.0-develop-680b6bb50c2b41af59cec25a49afd9b2cb98de10
v1.9.0-develop-680b6bb50c2b41af59cec25a49afd9b2cb98de10
```

```shell
~/eos $ echo "//yo" >> programs/cleos/main.cpp;
~/eos $ git status
On branch version-library
Your branch is ahead of 'origin/version-library' by 2 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   programs/cleos/main.cpp
```

```shell
~/eos $ pushd ./; cd ninja; ninja; popd; pushd ./; cd make; make -j20; popd;
.
.
.
```

```shell
~/eos $ ./ninja/bin/cleos version full; ./make/bin/cleos version full;
v1.9.0-develop-98f88862f3ba286ee3780dbb89175b5436432537
v1.9.0-develop-98f88862f3ba286ee3780dbb89175b5436432537-dirty
```

```shell
~/eos $ pushd ./; cd ninja; ninja; popd; pushd ./; cd make; make -j20; popd;
.
.
.
```

```shell
~/eos $ ./ninja/bin/cleos version full; ./make/bin/cleos version full;
v1.9.0-develop-98f88862f3ba286ee3780dbb89175b5436432537-dirty
v1.9.0-develop-98f88862f3ba286ee3780dbb89175b5436432537-dirty
```

**Comment by johndebord** [comment]
@spoonincode Fixed. Turns out that `ninja` aggressively starts building before `version_impl.cpp` even has a chance to get generated (therefore `ninja` never sees that `version_impl.cpp` has been touched). 

This is why the `version` library would only get built on the second iteration of using `ninja`; because on the second iteration, `ninja` sees that `version_impl.cpp` has been touched.

**Pull request merged**